### DISCUSSION TITLE
What does the “imageSizes” configuration option for <Image /> do?

### DISCUSSION THREAD

**Discussion by grazianodev** [vercel/next.js#24771] [Help]
I have been using Next.js new optimized Image component, and I think I understand how it works when using the different layout options (intrinsic, fixed, responsive and fill). However there is a configuration option called imageSizes that I don't understand.

From the docs, I can tell there are these two default configuration options:

```
module.exports = {
  images: {
    deviceSizes: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
    imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
  },
}
```

If I understand correctly, devicesSizes is an array of breakpoints of sorts; when using the fill or responsive layout for an image, Nextjs creates multiple versions of the image whose widths correspond to the breakpoints, and references those versions in the srcset that it automatically generates for that image. We may also tell the browser how to use these versions via the sizes prop. For example:

```
<Image 
    image={ image }
    layout="responsive" 
    sizes="(max-width: 500px) 100vw, 25vw" 
    width={1200}
    height={960}
/>
```

For imageSizes, the docs say that:

> You can specify a list of image widths using the imageSizes property. These widths should be different (usually smaller) than the widths defined in deviceSizes because the arrays will be concatenated. These widths are used when the next/image component uses layout="fixed" or layout="intrinsic".

Does this mean that, when using a fixed or intrinsic layout, Nextjs creates multiple versions of the image whose widths correspond to the values in imageSizes? If so, how can you tell the <Image /> component which one to use (since the docs also say that the size prop should be used with the fill and responsive layouts only)?

In other words, if I have an image that is 1200px wide and I want to use it at one of the sizes specified in imageSizes, how do I do that?

**Comment by bobarros** [comment]
You don't choose. The algorithm it's going to choose, based on the scenario.

The first thing that you have to realize is that one group (deviceSizes) is for large images and two layouts: fill and responsive:
The other is for small images (imageSizes) and the last two layouts: fixed and instrisic.

Let's run some test, and see the results:

Next config:

```json
images: {
    deviceSizes: [640, 750, 828, 1080, 1200],
    imageSizes:  [32,  64,  96,  128,   256],
}
```

Component config:
```jsx
const ImageUse = () => {
  return (
    <>
      <div style={{ position: 'relative', display: 'block', width: '1100px', height: "766px" }}>
        <Image layout="fill" src="/assets/image_1.jpg" />
      </div>
      <div style={{ position: 'relative', display: 'block', width: '100%', height: "auto" }}>
        <Image
          layout="responsive"
          src="/assets/image_2.jpg"
          width="700"
          height="560"
        />
      </div>
      <div style={{ position: 'relative', display: 'block', width: '200px', height: "160px" }}>
        <Image
          layout="fixed"
          src="/assets/image_3.jpg"
          width="180"
          height="64"
        />
      </div>
      <div style={{ position: 'relative', display: 'block', width: '100px', height: "72px" }}>
        <Image
          layout="intrinsic"
          src="/assets/image_4.jpg"
          width="96"
          height="60"
        />
      </div>
    </>
  );
};
```

This is the result:

```txt
originImage: 1475w x 1180h (jpg)

viewport:    1100w
1 and 2: 1200w x 960h 
3 and 4: 250w x 205h

viewport:    900w
1 and 2: 1080w x 864h 
3: 256w x 205h
4: 96w x 77h

viewport:    700w
1 and 2: 640w x 512h 
3: 256w x 205h
4: 96w x 77h
```

Do you see? Based on the range that the image fall with the correspondent config, you get a different result.

It's not quite easy to guess which choice it's going to be. The 900w viewport got and 1080w image (the choice above viewport size). The 700 had a smaller image than the viewport.

You don't precisely choose the image, but that doesn't mean that you don't have a lot of control.

**Comment by grazianodev** [comment]
Thank you, that helped. For some reason I was convinced the width and height set on <Image /> had to be the image's natural width and height; by playing with your example I realized that's not the case. 

My question originated from a use case I have where I can only seem to achieve my goal (a strip of photo thumbnails that retain the image's ratio when all reduced to the same height) with a fixed or intrinsic layout, so I have big images that I want to use at a small size. By setting small sizes on the <Image /> component I was able to get one of the images generated as per imageSizes.
